pandas/core/internals.py

```diff
@@ -2905,6 +2905,34 @@ def shift(self, periods, axis=0, mgr=None):
         return [self.make_block_same_class(new_values,
                                            placement=self.mgr_locs)]
 
+    def diff(self, n, axis=0, mgr=None):
```

Comment:
this should be in DatetimeBlock, puzzled why this works

Reply:
You had mentioned, https://github.com/pandas-dev/pandas/pull/19773#discussion_r169189270, that this should belong in DatetimeTZBlock. 

Additionally, DatetimeBlock can handle 2D data which can use the `algos.diff` method (dispatched via `Block.apply('diff', **kwargs)`) since that data can be converted to np.datetime64 cleanly. 

`DatetimeTZBlock` here overwrites the Block diff method that calls `algos.diff` but mimics the signature so everything is happy.

Reply:
yeah, ok for now. this whole diff bizness needs to be completely refactored.